### PR TITLE
docs: Overhaul the documentation for SIL and add a new SIL printing mode

### PR DESCRIPTION
**Pull request by eeckstein**
The main changes in the documentation are:

* Created a new directory `docs/SIL` and  moved all SIL-related files into this directory.

* Converted `rst` files to markdown.

* Extracted sections from `SIL.md` which go into very much detail - including the instruction reference - into separate files: `Types.md`, `Ownership.md`, `FunctionAttributes.md`, `Instructions.md`. Those files are referenced from `SIL.md` at the relevant places.

* Rewrote and updated the OSSA part in `SIL.md`

* Removed a few sections, which are not relevant anymore, like "Value dependency" (which is replaced by ownership concepts).

* Fixed and improved a lot of small things.

### CODE REVIEW

**Review by gottesmm — COMMENTED** [review]
Just some small flow comments.

File: docs/SIL/SIL.md [hunk 1]
```diff
@@ -0,0 +1,1536 @@
...
+## Ownership
+
+> **_Note:_** Many more details of OSSA are described in the [Ownership](Ownership.md) document.
+
+In OSSA each non-trivial SIL value is statically mapped to an _ownership kind_:
+
+- `@owned`: A value that exists independently of any other value and is
+  consumed exactly once along all control flow paths through a function.
+  "Consuming" the value means that it is either destroyed (via `destroy_value`)
+  or consumed by a consuming instruction (e.g. stored to memory with `store`).
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply docs/SIL/SIL.md:354]: https://en.wiktionary.org/wiki/e.g.

[the remark this answers is not in the document]
Comment (atrick) [reply docs/SIL/SIL.md:354]: I'm alarmed to see `e.x.` anywhere in our docs. what is that?

**Review by atrick — APPROVED** [review]
💯 on
- -emit-pure-sil
- documentation structure
- OSSA examples (they look familiar, good to have them in the docs)

I don't think we need to worry about typos here or there in this PR

File: docs/SIL/SIL.md [hunk 2]
```diff
@@ -0,0 +1,1536 @@
...
+The ownership kind of a value is statically determined:
+
+- Basic block arguments have their ownership explicitly specified: 
+
+```
+    bb1(%0 : @owned $String, %1 : @guaranteed $String, %2 : $Int):
+      ...
+```
+
+- The ownership of instruction results depends on the kind of instruction. Most
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply docs/SIL/SIL.md]: I reworded it a bit

File: docs/SIL/SIL.md [hunk 3]
```diff
@@ -0,0 +1,1536 @@
...
+  produces an owned value, whereas `begin_borrow` always produces a guaranteed
+  value.
+
+- Forwarding instructions: some instructions work with both, owned and
+  guaranteed ownership, and "forward" the ownership from their operand(s) to
+  their result(s), for example cast instructions.  
+  If a forwarding instruction has multiple operands (like `struct` or `tuple`),
+  the ownership of all operand values must be consistent. The operands values
+  cannot have both guaranteed and owned ownership. However, it's possible to
+  mix operands with ownership with trivial operands.
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply docs/SIL/SIL.md:420]: As you said, that's in Ownership.md now

File: docs/SIL/SIL.md [hunk 4]
```diff
@@ -0,0 +1,1536 @@
...
+```
+      %1 = load [copy] %0   // producer
+      %2 = copy_value %1    // interior use of %1
+      %3 = move_value %1    // consuming use of %2
+      %4 = copy_value %1    // ERROR: use of %1 outside %1's lifetime
+```
+
+- Forwarding instructions end owned lifetimes and immediately produce a new
+  owned value, which introduces a separate lifetime. The combined lifetimes over
+  all forwarding instructions is called _forward-extended_ lifetime.
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply docs/SIL/SIL.md]: A forward-extended lifetime does not refer to a single value. It involves multiple values, even multiple definitions (producers).
I think this is important to know and I added a paragraph+example for this.

**Comment by rjmccall** [comment]
I love the SIL documentation restructuring, that seems great.

Having a SIL printing mode that suppresses all the theoretically-helpful comments feels like a totally reasonable feature, albeit one that should be added in a separate PR from the docs rework.  I don't like the idea of having multiple dialects of SIL itself, though — if we agree that this is a better way to write SIL, we should just change SIL.  (We can probably write a script to mass-update the tests.)

The idea of removing types from printed operands is interesting. A lot of instructions already avoid operand types when they're redundant with other information in the instruction, and that's something we could pretty easily just do more of without dramatically changing how we parse SIL. If we want to do it in general by propagating type information from the value definition, we're going to run into the problem that basic blocks are not required to appear in an order consistent with dominance — SILGen usually (always?) emits them this way, but CFG optimizations can definitely mess that up.  The general solution to that problem is to introduce a parse tree, although I suppose we could also just recognize that we're processing out of order and skip tokens until we reach the next block. Or we could make operand types *optional* and then only print them when blocks are rendered out of order?

**Comment by eeckstein** [comment]
> one that should be added in a separate PR from the docs rework.

Yes, that makes sense. I removed the SIL printing commit and will put it into a separate PR.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Pull request merged**